doc/rst/technotes/partialInstantiations.rst

```diff
@@ -0,0 +1,250 @@
+
+.. code-block:: chapel
+
+  // dependent.chpl
+  record R {
+    type T;
+    param p : T;
+  }
+
+  type A = R(p=5);
+
+Output:
+
+.. code-block:: text
+
+  dependent.chpl:7: error: Unable to resolve partial instantiation 'R(p=5)'
+  dependent.chpl:7: note: Instantiation of field 'p' depends on uninstantiated fields: 'type T'
+
+Generic fields may have default values that are used to instantiate the field
+when the user does not explicitly provide an argument in the type constructor.
+In order to allow these fields to remain generic the user can pass ``?`` as an
+argument to the type constructor. This argument indicates that generic fields
+with default values should not be automatically instantiated.
+
+.. code-block:: chapel
+
+  record MyRange {
+    type idxType = int;
+    param stridable = false;
+  }
+
+  // all arguments must be of type ``MyRange(int,false)``
+  proc foo(arg : MyRange) { ... }
+
+  // Any instantiation of ``MyRange`` is accepted
+  proc bar(arg : MyRange(?)) { ... }
+
+  // Any ``MyRange`` accepted provided the first field is ``uint``
+  proc baz(arg : MyRange(uint, ?)) { ... }
+
+There are some restrictions that accompany the ``?`` argument. First is that
+``?`` may only appear once in a type constructor call. A user might be
+tempted to use ``?`` as a positional argument in order to more easily pass
+an argument to a specific field, e.g. ``R(?, ?, int)``. In order to reduce
+confusion and support ``?`` as a convenient catch-all argument, ``?`` may not
+be used multiple times in a type constructor call. For example:
+
+.. code-block:: chapel
+
+  record R {
+    type X = int;
+    type Y = real;
+    type Z = string;
+  }
+
+  // Error! '?' may not be used multiple times in a type constructor call
+  // proc foo(arg : R(?, ?, uint)) { ... }
+  // instead, use named-expressions:
+  proc foo(arg : R(?, Z=uint)) { ... } // or 'R(Z=uint, ?)'
+
+Furthermore, once ``?`` appears in a type constructor call all later arguments
+in the same call must used named expressions. This restriction emphasizes that
+``?`` may not be used as a positional argument. For example:
+
+.. code-block:: chapel
+
+  record R {
+    type A = int;
+    type B = real;
+  }
+
+  // Error! second argument must use a named expression
+  // proc foo(arg : R(?, string)) { ... }
+  // instead...
+  proc foo(arg : R(?, B=string)) { ... }
+
+Detecting Partial Instantiations
+--------------------------------
+
+There are a few features that allow for the detection of partial
+instantiations.
+
+The first feature is a standard library function named ``isGenericType`` that
+will simply return ``true`` if its argument is a partial instantiation or a
+fully-generic type:
+
+.. code-block:: chapel
+
+  record R {
+    type T;
+    param p : int;
+  }
+
+  writeln(isGenericType(R));            // true
+  writeln(isGenericType(R(int)));       // true
+  writeln(isGenericType(R(p=42)));      // true
+  writeln(isGenericType(R(string, 5))); // false
```

Comment:
I would probably document `isGeneric(R)` instead because I don't see the point in having `isBlaType` and `isBlaValue` and am hoping we deprecate / make not user facing the Type and Value versions.